### DISCUSSION TITLE
Is this the best way to peform dimensionality reduction with PCA in dlib?

### DISCUSSION THREAD

**Discussion by arrufat** [davisking/dlib#2976] [Q&A] [answered]
I noticed dlib has a `vector_normalizer_pca` that performs dimensionality reduction based on energy.
However, I want to output to a fixed number of dimensions.  For that, I compute the PCA with 100% energy (so no dimensions are lost) and then take a range from that PCA matrix.

I've been using this approach for some time, but I wonder if there is a simpler way?

```c++
using namespace dlib;
auto project(const std::vector<matrix<float, 0, 1>>& embeddings, const size_t output_dims) -> std::vector<matrix<float, 0, 1>>
{
    vector_normalizer_pca<matrix<float, 0, 1>> normalizer;
    const matrix<float, 0, 0> pca = normalizer.pca_matrix();
    normalizer.train(embeddings, 1); 
    const auto& m = normalizer.means();
    const auto& s = normalizer.std_devs();
    const matrix<float, 0, 0> M = rowm(pca, range(0, output_dims - 1));
    std::vector<matrix<float, 0, 1>> projected;
    for (size_t i = 0; i < embeddings.size(); ++i)
    {
        projected.push_back(M * pointwise_multiply(embeddings[i] - m, s));
    }
    return projected;
}
```

**Comment by davisking** [comment, marked as answer]
You have 
```
    const matrix<float, 0, 0> pca = normalizer.pca_matrix();
    normalizer.train(embeddings, 1);
```
backwards, but yeah that's all there is in dlib. In the past whenever I was doing this stuff it was at work and just kinda different each time.  Depends on the context.  Like often you just call svd() directly.

**Reply by arrufat** [reply]
Ah! I just refactored the code I had in here directly...

Yes, I get how sometimes you care about the energy "compression" and some other times about the dimensionality reduction.

Right, I could just call `svd()`, too.

Thank you!
